### PR TITLE
Remove AS-NS linking in the AS

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

References #2312 

#### Changes

- Add `NsAs` service that allows NS to send uplinks to AS
- Relax `ApplicationInvalidatedDownlinks` validation, since we do use empty queues to signal `FCnt` changes
- Remove AS linking mechanism completely
- Retry Server removed (as links are no longer cancelled on update)
- `RxMetadata` and `TxSettings.DataRate.Modulation` added to AS tests, as we now go through the gRPC validation middleware, and these are required in order to actually make the calls
- Frontends can now choose if they subscribe only to locally processed traffic, or the cluster-wide traffic. Cluster traffic is routed over Pub/Sub (over Redis Pub/Sub in this PR), while local traffic is simply broadcasted/routed on send

#### Testing

As the NS part is not ready, I've adapted the `mockNS` to simulate a real NS, and the tests pass (but not much integration testing can be done without the NS).

##### Regressions

No behavioral changes are expected, but tests have been adapted to the new environment.

#### Notes for Reviewers

This is the initial groundwork for the AS-side removal. There are still some `TODOs` left that we should discuss, mainly about how to handle depreciation - I've added these in #2312 .

@johanstokking this is ready for an initial review.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/applicationserver.proto [hunk 1]
```diff
@@ -106,6 +106,15 @@ service As {
   };
 }
 
+message NsAsHandleUplinkRequest {
```
Comment (johanstokking) [inline api/applicationserver.proto:117]: I think that `ApplicationUps` is more consistent (i.e. we have `Clients` which has `repeated Client`)

File: api/applicationserver.proto [hunk 2]
```diff
@@ -106,6 +106,15 @@ service As {
...
+  repeated ApplicationUp application_uplinks = 1 [(validate.rules).repeated.min_items = 1];
```
Comment (johanstokking) [inline api/applicationserver.proto]: `application_ups`

File: pkg/applicationserver/applicationserver.go [hunk 3]
```diff
@@ -253,47 +238,69 @@ func (as *ApplicationServer) Roles() []ttnpb.ClusterRole {
...
+
+	return nil
 }
 
-func skipPayloadCrypto(link *link, dev *ttnpb.EndDevice) bool {
+func (as *ApplicationServer) broadcastUp(ctx context.Context, up *ttnpb.ApplicationUp) error {
+	if err := as.localDistributor.SendUp(ctx, up); err != nil {
+		return err
+	}
+	return as.clusterDistributor.SendUp(ctx, up)
```
Comment (johanstokking) [inline pkg/applicationserver/applicationserver.go]: Are they sequential for a reason? Are these calls blocking? If so, consider goroutines and a wait group

File: pkg/applicationserver/distribution/distribution.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// PubSub sends upstream traffic from publishers to subscribers.
+type PubSub interface {
+	// SendUp publishes traffic to the subscribers.
+	SendUp(context.Context, *ttnpb.ApplicationUp) error
```
Comment (johanstokking) [inline pkg/applicationserver/distribution/distribution.go]: I would assume this is blocking because of the name and the fact it returns an error. So that's ok, but please do this concurrent then

Comment (johanstokking) [reply pkg/applicationserver/distribution/distribution.go]: You might want to rename these to `Publish()` so that they match better with their `Subscribe()` counterparts and it aligns better with their godoc

File: pkg/applicationserver/distribution/pubsub_distributor.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	pubsub        PubSub
+	subscriptions *subscriptionMap
+}
+
+// SendUp sends traffic to the underlying Pub/Sub.
+func (d *pubSubDistributor) SendUp(ctx context.Context, up *ttnpb.ApplicationUp) error {
+	return d.pubsub.SendUp(ctx, up)
+}
+
+var errMissingIdentifiers = errors.DefineUnimplemented("missing_identifiers", "subscriptions without identifiers are not supported")
```
Comment (johanstokking) [inline pkg/applicationserver/distribution/pubsub_distributor.go]: That's a failed precondition

**Review by johanstokking — COMMENTED** [review]
Great stuff, some minor comments only

Please append fixup commits so that subsequent reviews go a bit faster

File: api/applicationserver.proto (hunk 1 above)
Comment (adriansmares) [reply api/applicationserver.proto:117]: The reasoning behind the naming was that later we may want to add other fields to this request, so `*Request` fits this better in that case.

File: pkg/applicationserver/applicationserver.go (hunk 3 above)
Comment (adriansmares) [reply pkg/applicationserver/applicationserver.go]: Error handling is better if they are sequential, and both just write to a buffered channel, so I'm more inclined to leave it as is.

Edit: ~actually Redis publish is blocking, but that's ok since it's not going to throttle the local call (which is non-blocking)~.
Edit2: Actually both are blocking, but I'm still not sold on spawning extra goroutines just for this. I want to do this change when we can pinpoint this as a throttle point.

Comment (johanstokking) [reply pkg/applicationserver/applicationserver.go]: I'm okay with optimizing this later.

File: api/applicationserver.proto (hunk 1 above)
Comment (johanstokking) [reply api/applicationserver.proto:117]: We can do that later and stay wire compatible. But I'm fine with either

**Review by johanstokking — APPROVED** [review]

**Comment by htdvisser** [comment]
> Remove AS linking mechanism completely

Maybe I missed something, but I thought that we had to keep the old linking mechanism for supporting external Application Servers. Did we drop that?

File: cmd/ttn-lw-stack/commands/start.go [hunk 6]
```diff
@@ -238,6 +239,9 @@ var startCommand = &cobra.Command{
 			config.AS.Devices = &asredis.DeviceRegistry{
 				Redis: redis.New(config.Redis.WithNamespace("as", "devices")),
 			}
+			config.AS.Distribution.PubSub = &asdistribredis.PubSub{
+				Redis: redis.New(config.Redis.WithNamespace("as", "traffic")),
```
Comment (htdvisser) [inline cmd/ttn-lw-stack/commands/start.go]: Shouldn't this use the "cache" Redis, since we're just using it for pub/sub, and not for storage?

File: pkg/applicationserver/grpc.go [hunk 7]
```diff
@@ -37,64 +37,54 @@ func (as *ApplicationServer) SetLink(ctx context.Context, req *ttnpb.SetApplicat
...
-		log.FromContext(ctx).WithError(err).Warn("Failed to cancel link")
-	}
 	_, err := as.linkRegistry.Set(ctx, *ids, nil, func(link *ttnpb.ApplicationLink) (*ttnpb.ApplicationLink, []string, error) { return nil, nil, nil })
 	if err != nil {
 		return nil, err
 	}
 	return ttnpb.Empty, nil
 }
 
+var errUnimplemented = errors.DefineUnimplemented("unimplemented", "unimplemented")
```
Comment (htdvisser) [inline pkg/applicationserver/grpc.go]: This is too generic. The description could for example say something about linking not being necessary in this deployment.

File: sdk/js/generated/api-definition.json [hunk 8]
```diff
@@ -736,6 +742,7 @@
         "provider",
         "provider.mqtt",
         "provider.mqtt.client_id",
+        "provider.mqtt.headers",
```
Comment (htdvisser) [inline sdk/js/generated/api-definition.json]: This looks like an unrelated change

**Comment by adriansmares** [comment]
> > Remove AS linking mechanism completely
> 
> Maybe I missed something, but I thought that we had to keep the old linking mechanism for supporting external Application Servers. Did we drop that?

My understanding was that from the beginning external linking was abandoned as soon as we opened https://github.com/TheThingsNetwork/lorawan-stack/issues/2312 due to the related scalability issues it creates.

The newer mentality imo being that 'having your external AS' is achieved by disabling crypto at application/device level and operating these crypto operations yourself (Violet-like).

File: pkg/applicationserver/grpc.go (hunk 7 above)
Comment (adriansmares) [reply pkg/applicationserver/grpc.go]: Linking is not going to be available in any deployment, but I can mention that the method has been deprecated.

**Comment by johanstokking** [comment]
> The newer mentality imo being that 'having your external AS' is achieved by disabling crypto at application/device level and operating these crypto operations yourself

Yep, that's it
